services/cd-service/pkg/repository/transformer.go

```diff
@@ -581,6 +603,32 @@ func (c *CreateUndeployApplicationVersion) Transform(ctx context.Context, state
 	return fmt.Sprintf("created undeploy-version %d of '%v'\n%s", lastRelease+1, c.Application, result), changes, nil
 }
 
+func removeCommit(fs billy.Filesystem, commitID, application string) error {
+	couldNotRemoveAppFormat := "could not remove the application from the commit directory %w"
+	dir := commitApplicationDirectory(fs, commitID, application)
+	if err := fs.Remove(dir); err != nil {
+		return fmt.Errorf(couldNotRemoveAppFormat, err)
+	}
+	// check if there are no other services updated by this commit
+	// if there are none, start remove the entire branch of the commit
+	for {
+		dir = path.Dir(dir)
+		if dir == "commits" {
+			break
+		}
+		files, err := fs.ReadDir(dir)
+		if err != nil {
+			return fmt.Errorf(couldNotRemoveAppFormat, err)
+		}
+		if len(files) == 0 {
+			if err = fs.Remove(dir); err != nil {
+				return fmt.Errorf(couldNotRemoveAppFormat, err)
+			}
+		}
+	}
+	return nil
+}
```

Comment:
I'm not a fan of loops, unless they are tied to a counter, because it's hard to see if it could be an infinite loop.

How about this (pseudo code):

```suggestion
        deleteDirIfEmpty := func(dir) {
     		files, err := fs.ReadDir(dir)
		if err != nil {
			return fmt.Errorf(couldNotRemoveAppFormat, err)
		}
		if len(files) == 0 {
			if err = fs.Remove(dir); err != nil {
				return fmt.Errorf(couldNotRemoveAppFormat, err)
			}
		}
        }
        deleteDirIfEmpty(commitApplicationDirectory(...));
        deleteDirIfEmpty(commitDirectory(...));
        deleteDirIfEmpty(path.Dir(commitDirectory(...)));
	return nil
}
```

Reply:
I agree with this, now it's refactored.